WIP: Pass required non-string arguments at add_/init_ time.

litghost opened the pull request:
Other changes:
 - Add `finish_` calls
 - Fix some pyflakes lint

litghost committed:
WIP: Pass required non-string arguments at add_/init_ time.

Signed-off-by: Keith Rothman <537074+litghost@users.noreply.github.com>

duck2 commented on uxsdcxx/cpp.py:
```diff
@@ -182,13 +218,29 @@ def _gen_load_simple(t: UxsdSimple, input: str) -> str:
+	args = ["data"]
+	load_args = []
+	
+	for attr in complex_type_map[t.type.name].attrs:
```
Why not just `t.type.attrs`?

litghost replied:
Is that the same as what I did?

duck2 replied:
Yes, `t.type` gives the exact same `UxsdComplex` you get from the map.

litghost committed:
Fix assort bugs found during testing.

Signed-off-by: Keith Rothman <537074+litghost@users.noreply.github.com>

litghost committed:
Use attrs in type.

Signed-off-by: Keith Rothman <537074+litghost@users.noreply.github.com>

litghost committed:
Add example usage of new interface.

This XML to capnp converter should be auto-generated, but I was
experimenting with how to implement it first.

Signed-off-by: Keith Rothman <537074+litghost@users.noreply.github.com>

litghost commented:
In https://github.com/duck2/uxsdcxx/pull/29/commits/858566b5c8813a4b6d396d2ee68d2b80a95e4ff5 I've added a hand-written example consumer of the new interface, which converts the input (XML in the example) to the capnproto schema.  In the long term if we are using capnproto, this file should be auto-generated, but I wasn't confident when I started how to implement the capnproto output.

I'm going to start on a generator that walks the capnproto message and calls back into the interface.  After that I'll need a VPR binding of the interface to test relative performance under XML and capnproto.  @duck2 Do you have a working VPR binding of the interface?

duck2 commented:
Unfortunately no, I'm working on the writer interface side currently.

litghost committed:
Add generated capnp interface usage.

Signed-off-by: Keith Rothman <537074+litghost@users.noreply.github.com>

litghost committed:
Add capnp to capnp example.

Signed-off-by: Keith Rothman <537074+litghost@users.noreply.github.com>

duck2 commented:
Pushed the write interface changes. The base class [looks like this](https://gist.github.com/duck2/b77f2001c4e2fd20bb0695e2955afbd2) now. I think I should start on the VPR interface, since I can test it more easily.

litghost commented:
> Pushed the write interface changes. The base class [looks like this](https://gist.github.com/duck2/b77f2001c4e2fd20bb0695e2955afbd2) now. I think I should start on the VPR interface, since I can test it more easily.

Can you incorporate my required attribute changes?

litghost commented:
Oh, and I think the required attribute changes should only apply on the reader interfaces.  The write interfaces are much less important.

litghost commented:
Also I split the interface header out from the XML reading header, which is a good idea.

duck2 commented:
Well, I couldn't find a way to keep the PR open while merging the changes up to now.
